### ISSUE TITLE
A way to have mochify fail on error with gulp build task?

### ISSUE THREAD

**Issue by oller** [mantoni/mochify.js#109]
Hi there,

I've integrated mochify into my TDD workflow, however I'd also like to perform these tests as part of a gulp build task which will fail the build if tests fail.

I'm not sure if this is something that should be addressed by gulp or mochify, but I can't find an `failOnError` callback or something to that extent in mochify.  At the moment, despite calling the mochify task first on in my build task - from the console output it runs at the very end, even if it fails, so the build step is still completed.

The relevant gulp task:

```js
// Mochify Unit Testing
gulp.task('test', () => {
  mochify('./test/*.js', {
    reporter: 'landing',
    cover : true
  }).bundle();
});
```

The generated console output
```bash
→ gulp build
[10:07:42] Requiring external module babel-core/register
[10:07:43] Using gulpfile ~/Playground/gulp-browserify-backbone-mdl-webapp/gulpfile.babel.js
[10:07:43] Starting 'test'...
[10:07:43] Finished 'test' after 25 ms
[10:07:43] Starting 'lint'...
[10:07:44] Starting 'styles'...
[10:07:44] Starting 'browserify'...
[10:07:44] Starting 'images'...
[10:07:45] Starting 'fonts'...
[10:07:45] Starting 'extras'...
[10:07:45] Finished 'images' after 653 ms
[10:07:45] Finished 'fonts' after 68 ms
[10:07:46] Finished 'extras' after 1.2 s
[10:07:46] Finished 'styles' after 2.64 s
[10:07:46] Finished 'lint' after 3.03 s
[10:07:48] Finished 'browserify' after 3.71 s
[10:07:48] Starting 'html'...
[10:07:53] Finished 'html' after 5.38 s
[10:07:53] Starting 'build'...
# phantomjs:
[10:07:53] 'build' all files 77.83 kB (gzipped)
[10:07:53] Finished 'build' after 282 ms

  --------------------------------------------------------------------------------------------------------------------------------------
  ⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅✈
  --------------------------------------------------------------------------------------------------------------------------------------

  0 passing (9ms)
  1 failing
```

Any help appreciated. I'm a little new to these tools, so apologies if it's a daft question.

**Comment by mantoni** [comment]
Is this maybe the same issue as <https://github.com/mantoni/mochify.js/issues/108>?

**Comment by oller** [comment]
Certainly related, using the gulp pattern described in that ticket I can get closer, but to have it run as a precursor to the build and fail the build I have to steer away from gulp's nature of running things concurrently.  I'll keep on tinkering and update the issue if I find a solution.

As an aside - I'm using 'hbsfy' for templating, if I add this as a flag into mochify the template rendering works correctly allowing for testing however it breaks the istanbul coverage plugin.
```js
// Mochify Unit Testing
gulp.task('test', () => {
  mochify('./test/*.js', {
    reporter: 'spec',
    // cover : true,
    watch: true,
    transform: 'hbsfy'
  })
  //.plugin('mochify-istanbul', {
  //  report: 'html',
  //  dir: './test/coverage'
  //})
  .bundle();
});
```

Fails with:

```bash
TypeError: Cannot read property 'text' of undefined
    at /node_modules/mochify-istanbul/node_modules/istanbul/lib/report/html.js:283:53
    at Array.forEach (native)
    at annotateBranches (/node_modules/mochify-istanbul/node_modules/istanbul/lib/report/html.js:250:30)
    at HtmlReport.Report.mix.writeDetailPage (/node_modules/mochify-istanbul/node_modules/istanbul/lib/report/html.js:421:9)
    at /node_modules/mochify-istanbul/node_modules/istanbul/lib/report/html.js:484:26
```

Appoligies if this has been asked before, but is there a way to scope the transform just for the mochify part?

**Comment by mantoni** [comment]
You have to use mochify-istanbul's  `--exclude` option to exclude the templates from instrumentation. I'm doing this for mustache templates.

**Comment by mantoni** [comment]
@oller Did my suggestion help? Can this be closed?

**Comment by mantoni** [comment]
If this is still an issue, I can reopen. Closing for now.

**Comment by MarkVaughn** [comment]
Same issue, travis ci passes because gulp exits with 0
> Finished 'test' after 27 s
> Error: Exit 1
> The command "gulp test" exited with 0.

```
return mochify('./src/js/__tests__/**/*.js', {
        extension: ['.jsx', '.js'],
        node: true,
        path: ['./src/js/'],
        watch: true,
    })
        .transform(babelify)
        .transform(envify({
            _: 'purge',
            NODE_ENV: 'production',
        }))
        .bundle();
```

**Comment by mantoni** [comment]
@MarkVaughn You passed `watch: true`. I'm surprised it exited at all.

**Comment by mantoni** [comment]
Maybe a solution to this (reoccurring) issue could be to set `process.exitStatus = 1` in any error cases, regardless of watch mode, and remove the explicit `process.exit()` calls. That would make shutdown a lot cleaner and not exit unintentionally when used via the API.

**Comment by MarkVaughn** [comment]
you are right, I had two tasks, I copied the wrong one, travis used  the same thing with watch false lol.
but here's what I got in order for it to make travis fail:
```
return mochify('./src/js/__tests__/**/*.js', {
        extension: ['.jsx', '.js'],
        node: true,
        path: ['./src/js/'],
        watch: false,
    })
        .transform(babelify)
        .transform(envify({
            _: 'purge',
            NODE_ENV: 'production',
        }))
        .on('error', function(){
            process.exit(1);
        }).bundle()
```

**Comment by mantoni** [comment]
@MarkVaughn Thanks for the update. Reopening.

**Comment by mantoni** [comment]
I just released `2.18.1` with a fix for this. Please confirm that this works fine now for you, otherwise I'll reopen (again). Thanks for your patience.
